fees/estimator.go

```diff
@@ -0,0 +1,902 @@
+	if stats.db == nil {
+		return errors.New("estimator database is not open")
+	}
+
+	// Database version is currently hardcoded here as this is the only
+	// place that uses it.
+	currentDbVersion := []byte{1}
+
+	version, err := stats.db.Get(dbKeyVersion, nil)
+	if err != nil && err != leveldb.ErrNotFound {
+		return fmt.Errorf("error reading version from db: %v", err)
+	}
+	if len(version) < 1 {
+		// No data in the file. Fill with the current config.
+		batch := new(leveldb.Batch)
+		b := bytes.NewBuffer(nil)
+		maxConfirmsBytes := make([]byte, 4)
+		bestHeightBytes := make([]byte, 8)
+
+		batch.Put(dbKeyVersion, currentDbVersion)
+
+		dbByteOrder.PutUint32(maxConfirmsBytes, uint32(stats.maxConfirms))
+		batch.Put(dbKeyMaxConfirms, maxConfirmsBytes)
+
+		dbByteOrder.PutUint64(bestHeightBytes, uint64(stats.bestHeight))
+		batch.Put(dbKeyBestHeight, bestHeightBytes)
+
+		err := binary.Write(b, dbByteOrder, stats.bucketFeeBounds)
+		if err != nil {
+			return fmt.Errorf("error writing bucket fees to db: %v", err)
+		}
+		batch.Put(dbKeyBucketFees, b.Bytes())
+
+		err = stats.db.Write(batch, nil)
+		if err != nil {
+			return fmt.Errorf("error writing initial estimator db file: %v",
+				err)
+		}
+
+		err = stats.updateDatabase()
+		if err != nil {
+			return fmt.Errorf("error adding initial estimator data to db: %v",
+				err)
+		}
+
+		log.Debug("Initialized fee estimator database")
+
+		return nil
+	}
+
+	if !bytes.Equal(currentDbVersion, version) {
+		return fmt.Errorf("incompatible database version: %d", version)
+	}
+
+	maxConfirmsBytes, err := stats.db.Get(dbKeyMaxConfirms, nil)
+	if err != nil {
+		return fmt.Errorf("error reading max confirmation range from db file: "+
+			"%v", err)
+	}
+	if len(maxConfirmsBytes) != 4 {
+		return errors.New("wrong number of bytes in stored maxConfirms")
+	}
+	fileMaxConfirms := int32(dbByteOrder.Uint32(maxConfirmsBytes))
+
+	feesBytes, err := stats.db.Get(dbKeyBucketFees, nil)
+	if err != nil {
+		return fmt.Errorf("error reading fee bounds from db file: %v", err)
+	}
+	if feesBytes == nil {
+		return errors.New("fee bounds not found in database file")
+	}
+	fileBucketFees := make([]feeRate, len(feesBytes)/8)
+	err = binary.Read(bytes.NewReader(feesBytes), dbByteOrder,
+		&fileBucketFees)
+	if err != nil {
+		return fmt.Errorf("error decoding file bucket fees: %v", err)
+	}
+
+	if !replaceBuckets {
+		if stats.maxConfirms != fileMaxConfirms {
+			return errors.New("max confirmation range in database file different " +
+				"than currently configured max confirmation")
+		}
+
+		for i, f := range fileBucketFees {
+			if stats.bucketFeeBounds[i] != f {
+				return errors.New("bucket fee rates stored in database file " +
+					"different than currently configured fees")
+			}
+		}
+	}
+
+	fileBuckets := make([]txConfirmStatBucket, len(fileBucketFees))
+
+	iter := stats.db.NewIterator(ldbutil.BytesPrefix(dbKeyBucketPrefix), nil)
+	err = nil
+	fbytes := make([]byte, 8)
```

Comment:
This could be a stack allocation already regardless of whether it was created with `make`.  If it has a known length, is not too large, and passes escape analysis, the backing array will be stack allocated.

Reply:
Changed this and other instances.